### DISCUSSION TITLE
invariant

### DISCUSSION THREAD

**Discussion by cjquines** [remeda/remeda#963] [Suggest a Function]
From #469 . Could also be named `assert`, `throwIf`, etc.

I think, for `invariant` to be most useful in data-last cases, it should have the type

```ts
function invariant<T, U extends T>(
    guard: (data: T) => data is U
): (data: T) => U;
```

so you can call it like

```ts
pipe(
  data as number | string,
  invariant(isNumber),
  data => data, // data has type number 
);
```

But, I think for `invariant` to be useful for data-first cases, the type would be quite different

```ts
function invariant<T, U extends T>(
    data: T,
    guard: (data: T) => data is U
): asserts data is U;
```

as in

```ts
(data: number | string) => {
  invariant(data, isNumber);
  data // data has type number
}
```

Note that in the data-last case, we return `data`  unchanged; in the data-first case, we return `void`.

For non-guard cases, I don't think we can carry type-level information, so we could have an overload that takes a guard of `(data: T) => boolean` and doesn't narrow `data`.

See [playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEpwGZQiOAIigFMQSATZQgbgCg6B6RuKmZAWm2ClRjookAxgBtkpHAFcAdkJjAI0uMGkA3ccGTSYAHgAqAGjgBVOCQAeMEtIro9APgAUdOK9bJ2ALjiGXbgOaS4hTejmzI3noAlHAAvPbu7MroxnRR3mioJLDo4ckm9Ews4Vw8fEbSEHAwAJ5gJMrS2NAgHgrSAsJiEtgycu2N6lCa2vpOfq7hkQYTcIHBoVM+MfFwAEYQECIkWmneqhDAFIXMiZxifJ2i4g29svKKgxpauoYmZpbWtj7jbnNBUBCcDCHgiyziCTywBSe2BS2iEIKDFOJQuMAqVVq9UazSgrQeHUE1x6fQJT2GLzGzj+80Bi1BkRWCQ2Wx2HXScIZ4NWekKGAAdGBgPVqW4AIwodDSSQgNbZOAAHzgfGG0n8Mz+KiGIxgjgF0IAyjBVf4ohq3HlVuFza5TgA9AD8aUKIK8cGlsvlSpVKlNiIA3rMtc9tK7kEZ9agjSaovQ-uE4257U6AL5AA).

**Comment by eranhirsch** [comment]
I really want this! I've also been thinking about it recently. I think that `tiny-invariant` does some optimizations under-the-hood for production builds, which go beyond how we think about Remeda functions, and I don't want to suggest people stop using that library (and lose out on benefiting from those optimizations), but also don't want to just clone `tiny-invariant` inside Remeda as those kind of optimizations are what we usually consider; which brings me to think of having this as a utility that lives side-by-side with the regular invariant... 

In one of my project's I ended up calling it "validate" (so it doesn't name clash with `invariant`) and used tiny-invariant in the implementation. This might mean adding tiny-invariant as a dependency, but that would be a first runtime dependency for remeda, could also be a peer dependency I guess...

TL;DR - 🤷🏼

**Comment by cjquines** [comment]
having it as a peer dep would be fine i think? we could attempt to import it, and if not use our own... but then i'd be concerned about how that works with bundling

**Comment by cjquines** [comment]
hm, it looks like tiny-invariant's prod build optimization has to do with messages? if we don't have any messages with invariant (as in my proposal above) then we don't have to do that

**Comment by eranhirsch** [comment]
But we should support messages, it's an important part of throwing an invariant...

**Comment by cjquines** [comment]
okay more thoughts:

* `tiny-invariant` only checks `process.env.NODE_ENV`, and i'm not convinced that it makes the build size smaller without some special work on the bundler side. in particular, this doesn't work on environments that don't set this
* the guaranteed way to get prod optimization with `tiny-invariant` is to [use a plugin based on the bundler](https://github.com/alexreardon/tiny-invariant?tab=readme-ov-file#dropping-your-message-for-kb-savings). all these plugins don't care as to the actual library being used, as long as they fulfill the contract of `invariant(condition, ...args)`
* some users of `tiny-invariant` want the error message in prod ( see https://github.com/alexreardon/tiny-invariant/pull/172 ) so i don't think we should default to removing it, even if we could

thus:

* i think we can write a data-first version of invariant, `invariant(condition, ...args)`, and point users to the plugins under `tiny-invariant` docs if they care about saving space. (not sure if it'll work with `R.invariant` but eh)
* a data-last version of invariant shouldn't be `invariant(...args)(condition)` because it'd get rewritten by these plugins. i still think it's helpful to have, but we should call it something else. maybe `guard` idk

**Comment by airjp73** [comment]
Is this something y'all are still interested in having? I just swung by to make a similar feature request after making my own pipeable invariant.

I'd be happy to make a PR, but it sounds like there are still some design considerations to hash out here.

FWIW, I've recently switched to the [Epic Web invariant](https://github.com/epicweb-dev/invariant) specifically because it _doesn't_ optimize away messages. For app code, I've always felt like that was too much of a micro-optimization that ends up hurting the quality of your error reporting.
